typescript/cosmos-sdk/src/clients/provider.ts

```diff
@@ -0,0 +1,757 @@
+
+      if (igp) {
+        return AltVM.HookType.INTERCHAIN_GAS_PAYMASTER;
+      }
+    } catch {
+      try {
+        const { merkle_tree_hook } =
+          await this.query.postDispatch.MerkleTreeHook({ id: req.hookAddress });
+
+        if (merkle_tree_hook) {
+          return AltVM.HookType.MERKLE_TREE;
+        }
+      } catch {
+        throw new Error(`Unknown Hook Type: ${req.hookAddress}`);
+      }
+    }
+
+    throw new Error(`Unknown Hook Type: ${req.hookAddress}`);
+  }
+
+  async getInterchainGasPaymasterHook(
+    req: AltVM.ReqGetInterchainGasPaymasterHook,
+  ): Promise<AltVM.ResGetInterchainGasPaymasterHook> {
+    const { igp } = await this.query.postDispatch.Igp({ id: req.hookAddress });
+    assert(igp, `found no igp for id ${req.hookAddress}`);
+
+    const { destination_gas_configs } =
+      await this.query.postDispatch.DestinationGasConfigs({
+        id: igp.id,
+      });
+
+    const configs: {
+      [domainId: string]: {
+        gasOracle: {
+          tokenExchangeRate: string;
+          gasPrice: string;
+        };
+        gasOverhead: string;
+      };
+    } = {};
+
+    for (const config of destination_gas_configs) {
+      configs[config.remote_domain] = {
+        gasOracle: {
+          tokenExchangeRate: config.gas_oracle?.token_exchange_rate ?? '0',
+          gasPrice: config.gas_oracle?.gas_price ?? '0',
+        },
+        gasOverhead: config.gas_overhead,
+      };
+    }
+
+    return {
+      address: igp.id,
+      owner: igp.owner,
+      destinationGasConfigs: configs,
+    };
+  }
+
+  async getMerkleTreeHook(
+    req: AltVM.ReqGetMerkleTreeHook,
+  ): Promise<AltVM.ResGetMerkleTreeHook> {
+    const { merkle_tree_hook } = await this.query.postDispatch.MerkleTreeHook({
+      id: req.hookAddress,
+    });
+    assert(
+      merkle_tree_hook,
+      `found no merkle tree hook for id ${req.hookAddress}`,
+    );
+
+    return {
+      address: merkle_tree_hook.id,
+    };
+  }
+
+  // ### QUERY WARP ###
+
+  async getToken(req: AltVM.ReqGetToken): Promise<AltVM.ResGetToken> {
+    const { token } = await this.query.warp.Token({
+      id: req.tokenAddress,
+    });
+    assert(token, `found no token for id ${req.tokenAddress}`);
+
+    let token_type: AltVM.TokenType;
+
+    switch (token.token_type) {
+      case warpTypes.HypTokenType.HYP_TOKEN_TYPE_COLLATERAL:
+        token_type = AltVM.TokenType.collateral;
+        break;
+      case warpTypes.HypTokenType.HYP_TOKEN_TYPE_SYNTHETIC:
+        token_type = AltVM.TokenType.synthetic;
+        break;
+      default:
+        throw new Error(
+          `Failed to determine token type for address ${req.tokenAddress}`,
+        );
+    }
+
+    return {
+      address: token.id,
+      owner: token.owner,
+      tokenType: token_type,
+      mailboxAddress: token.origin_mailbox,
+      ismAddress: token.ism_id,
+      originDenom: token.origin_denom,
+      name: '',
+      symbol: '',
+      description: '',
+      decimals: 0,
```

Comment:
Isn't this bad? how can we get the token metadata on a cosmos token? Imagine if we have some code that expects these field not to be empty and it fails because this data is not populated in the first place

Reply:
That's weird. What about the `DenomMetadata` in the bank module? It can accommodate the name, symbol, and description. And it has `DenumUnits` where the decimals are set (called "exponent") in Cosmos SDK.

Aren't we using any of this?

Reply:
This was a missing feature request in the Hyperlane Cosmos Module. It will be probably added in the future, so currently it is unfortunately not supported right now.